wellcomeml/ml/clustering.py

```diff
@@ -0,0 +1,321 @@
+
+    def set_params(self, params, from_parameter_grid=False):
+        """
+        Set parameters of clustering, reducer and vectorizer
+
+        Args:
+            params: A dictionary of parameters
+            from_parameter_grid: Whether the dictionary is in the form of
+            sklearn parameter grid (e.g. {'clustering__eps': 1}), or a two
+            level dictionary (e.g. {'clustering': {'eps': 1}}) (default).
+
+        Returns:
+            None
+
+        """
+        self.vectorizer.set_params(**params.get('vectorizer', {}))
+        self.reducer.set_params(**params.get('reducer', {}))
+        self.clustering.set_params(**params.get('clustering', {}))
+
+    def _find_keywords(self, X_transformed, n_kw=10):
+        if self.embedding != 'tf-idf':
+            raise ValueError(f'Cannot find keyword for '
+                             f'{self.embedding} vectorizer')
+        idx_to_word = {value: key for key, value in
+                       self.vectorizer.vectorizer.vocabulary_.items()}
+
+        # Initialise a dictionary to calculate the aggregated vector for each
+        # cluster
+
+        aggregated = {key: np.zeros(X_transformed.shape[1])
+                      for key in self.cluster_ids}
+
+        for i in range(len(X_transformed)):
+            aggregated[self.cluster_ids[i]] += X_transformed[i]
+
+        kw_dictionary = {
+            key: ','.join(_find_words_from_frequency(vector,
+                                                     idx_to_word,
+                                                     n_kw=n_kw))
+            for key, vector in aggregated.items()
+        }
+
+        self.cluster_kws = [
+            kw_dictionary.get(cluster_id) for cluster_id in self.cluster_ids
+        ]
+
+
+def _find_words_from_frequency(vector, idx_to_word, n_kw=5):
+    """
+    Given a vector of frequencies and a dictionary connecting indexes
+     to word, find most prominent words
+    up to n_kw, eliminating possible duplicates or sub-words.
+
+    Args:
+        vector: A vector representing frequencies (or any other scores) for
+        each word.
+        idx_to_word: A dictionary mapping indices in `vector` to words.
+        n_kw: number of "prominent" words to output.
+
+    Returns:
+        A list of n_kw words (strings).
+
+    """
+    sorted_idx = np.argsort(vector)[::-1]
+    # Tests if new words do not overlap with existing ones (this is to
+    # exclude things like: 'cluster,clustering,clustered')
+
+    current_kw_list = []
+    i = 0
+
+    while len(current_kw_list) < n_kw and i <= len(vector)-1:
+        new_word = idx_to_word[sorted_idx[i]]
+        used_word = False
+        for kw in current_kw_list:
+            if new_word in kw or kw in new_word:
+                used_word = True
+                break
+
+        if not used_word:
+            current_kw_list += [new_word]
+
+        i += 1
+
+    return current_kw_list
+
+
+def _clustering_score(estimator, X, y=None):
```

Comment:
it is worth noting that silhouete is only one of possible scores. coherence has been shown to be a better predictor of topic quality https://radimrehurek.com/gensim/models/coherencemodel.html#gensim.models.coherencemodel.CoherenceModel and https://mimno.infosci.cornell.edu/info6150/readings/N10-1012.pdf. It would be great if we add that now or design the code in such a way that coherence can easily be added in the future